### PR TITLE
Basic fixes to calibre-sync

### PR DESCRIPTION
**Pull request by ToxicFrog**
Specifically:
- wake up periodically when ZMQs are active to process them, even if there are timed events pending
- create intermediate directories when receiving books to directories that don't exist yet.

This is enough for basic usage; koreader will connect to calibre and can receive books. Other features like the list of books on the device being displayed in calibre do not yet work, however.

### CODE REVIEW

File: plugins/calibrecompanion.koplugin/main.lua [hunk 1]
```diff
@@ -315,11 +315,52 @@ function CalibreCompanion:sendBooklists(arg)
...
+
+local function path_exists(path)
+    return lfs.attributes(path, "mode") ~= nil
+end
+
+-- strip the trailing element from path
+-- note: dirname foo/bar and dirname foo/bar/ are both foo
+-- dirname / is /
+-- dirname foo is .
+local function dirname(path)
```
Comment (Frenzie) [inline plugins/calibrecompanion.koplugin/main.lua]: Ah, it's too bad you wrote this. We have some util functions for that.

http://koreader.rocks/doc/modules/util.html#splitFilePathName

File: plugins/calibrecompanion.koplugin/main.lua [hunk 2]
```diff
@@ -315,11 +315,52 @@ function CalibreCompanion:sendBooklists(arg)
...
+    path = normalize_path(path):gsub("[^/]+/?$", "")
+    if path == "" then
+        return "."
+    end
+    return path
+end
+
+-- as `mkdir -p`
+-- unlike lfs.mkdir(), does not error if the directory already exists.
+local function mkdirs(path)
```
Comment (Frenzie) [inline plugins/calibrecompanion.koplugin/main.lua]: This probably makes more sense in util. Maybe `util.makeDir()`?

File: plugins/calibrecompanion.koplugin/main.lua [hunk 3]
```diff
@@ -315,11 +315,52 @@ function CalibreCompanion:sendBooklists(arg)
...
+
+-- We make the simplifying assumption that no-one will want to store books with
+-- a literal \ in the filename.
+-- FIXME: can calibre-on-windows even send paths with \ in them? If not we don't
+-- need to worry about this.
+local function normalize_path(path)
+    return (path:gsub('\\', '/'):gsub('/+', '/'))
+end
+
+local function path_exists(path)
```
Comment (Frenzie) [inline plugins/calibrecompanion.koplugin/main.lua]: Nice to have as a general thing. Could you please add it to util?

https://github.com/koreader/koreader/blob/15ec47d7e019c40013c5440c62862a751e6f8733/frontend/util.lua#L291-L305

File: frontend/ui/uimanager.lua [hunk 4]
```diff
@@ -660,6 +663,31 @@ function UIManager:resetInputTimeout()
...
+    end
+end
+
+-- Process all pending events on all registered ZMQs.
+function UIManager:processZMQs()
+    for _, zeromq in ipairs(self._zeromqs) do
+        for input_event in zeromq.waitEvent,zeromq do
+            self:handleInputEvent(input_event)
+        end
+        -- input_event = zeromq:waitEvent()
```
Comment (Frenzie) [inline frontend/ui/uimanager.lua]: What's this commented out stuff for exactly? It looks like it should probably just be removed?

Comment (ToxicFrog) [reply frontend/ui/uimanager.lua]: Done.

File: plugins/calibrecompanion.koplugin/main.lua (hunk 3 above)
Comment (ToxicFrog) [reply plugins/calibrecompanion.koplugin/main.lua]: Moved to `util.pathExists()`

File: plugins/calibrecompanion.koplugin/main.lua (hunk 1 above)
Comment (ToxicFrog) [reply plugins/calibrecompanion.koplugin/main.lua]: No worries, I had these functions lying around already, so it wasn't like I put a lot of time into it. :)

File: plugins/calibrecompanion.koplugin/main.lua (hunk 2 above)
Comment (ToxicFrog) [reply plugins/calibrecompanion.koplugin/main.lua]: Moved to `util.makeDirs()`

**Comment by ToxicFrog** [comment]
Made the requested changes but I haven't had a chance to test it on actual hardware yet, so even if it looks good I'd hold off on merging until I've had a chance to do that, hopefully tomorrow.

File: frontend/util.lua [hunk 5]
```diff
@@ -304,6 +304,31 @@ function util.isEmptyDir(path)
...
+    local lfs = require("libs/libkoreader-lfs")
+    return lfs.attributes(path, "mode") ~= nil
+end
+
+--- As `mkdir -p`.
+--- Unlike lfs.mkdir(), does not error if the directory already exists, and
+--- creates intermediate directories as needed.
+---- @string path the directory to create
+---- @treturn bool true on success; nil,err_message on error
+function util.makeDirs(path)
```
Comment (Frenzie) [inline frontend/util.lua]: Shall we get rid of the -s? It'd make me think you can pass a table.

Comment (ToxicFrog) [reply frontend/util.lua]: I stole the name from [python's os.makedirs function](https://docs.python.org/2/library/os.html#os.makedirs), where the -s differentiates between this function (which can create intermediate directories and ignores existing ones) and `mkdir()` (which does neither).

I'm not personally invested in the `makeDirs` name specifically, but I would like something to differentiate it from lower-level "create a single directory" functions like `lfs.mkdir`.

Comment (Frenzie) [reply frontend/util.lua]: Imo that's exactly what makeDir (as opposed to mkdir) does. ;-)

Anyway, wouldn't some variation on make*Path*/mkpath be a much clearer name then?

**Comment by ToxicFrog** [comment]
So, while testing this I discovered that `util.splitFilePathName("/foo/bar/")` returns `"/foo/bar/", ""`, not `"/foo/", "bar"`. I *thought* this was a bug (it's definitely unexpected), so I wrote a patch for it, but it looks like that there's a test that specifically checks this behaviour.

Is it safe to change this behaviour to match `dirname` (and update the test accordingly), or should I instead change my call sites to strip the trailing slash before calling?

**Comment by Frenzie** [comment]
The behavior of the function looks completely correct to me since `/foo/bar/` is a path without a file. Why is this unexpected?

**Comment by ToxicFrog** [comment]
I was expecting the behaviour to match that of POSIX `dirname`/`basename`, which don't care if the last element of the path you pass to it is a directory or a file; they just return "the directory containing the thing" and "the name of the thing". It looks like it actually matches bash `os.path.split`, which explicitly calls out that passing in a path with a trailing `/` will return `""` as the file-part.

I'll update the documentation to make this explicit and make no other changes to that function.

**Comment by ToxicFrog** [comment]
Done.

**Comment by Frenzie** [comment]
> I was expecting the behaviour to match that of POSIX dirname/basename

It's not my favorite function name, but the words split and file don't really support that interpretation for me. In fact I'd say they completely contradict it. :-P

**Comment by Frenzie** [comment]
Thanks!

**Comment by ToxicFrog** [comment]
You're welcome!

I hope to get file listing working next, but as you may have noticed I don't have a lot of free time to work on this lately. :)

**Pull request merged**